src/repo.rs

```diff
@@ -569,14 +567,58 @@ pub(crate) struct DocumentInfo {
     /// Ref count for handles(shared with handles).
     handle_count: Arc<AtomicUsize>,
     /// Per repo automerge sync state.
-    sync_states: HashMap<RepoId, SyncState>,
+    peer_connections: HashMap<RepoId, PeerConnection>,
     /// Used to resolve futures for DocHandle::changed.
     change_observers: Vec<RepoFutureResolver<Result<(), RepoError>>>,
     /// Counter of patches since last save,
     /// used to make decisions about full or incemental saves.
     patches_since_last_save: usize,
 }
 
+/// A state machine representing a connection between a remote repo and a particular document
+#[derive(Debug)]
+enum PeerConnection {
+    /// we've accepted the peer and are syncing with them
+    Accepted(SyncState),
+    /// We're waiting for a response from the share policy
+    PendingAuth { received_messages: Vec<SyncMessage> },
```

Comment:
For clarity, why not put the received messages on `ShareType::Synchronize`, since that seems to be the only use case of them? (that way it's clearer from the data, as opposed to having to understand the conditional logic)